File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticAlias.java
package com.foundgine.core.semantic;

import java.util.Objects;

/** Historical semantic name resolving to the owning canonical declaration. */
public record SemanticAlias(String name, Integer weight) {
    public SemanticAlias {
        Objects.requireNonNull(name);
        if (weight != null && (weight < 1 || weight > 100)) throw new IllegalArgumentException("Alias weight must be between 1 and 100 (inclusive) when specified.");
    }
    public SemanticAlias(String name) { this(name, null); }
    @Override public String toString() { return name; }
}

File: src/csharp/Foundgine.Core/Semantic/SemanticAlias.cs
namespace Foundgine.Core.Semantic;

/// <summary>
/// A historical semantic name that resolves to the owning canonical declaration.
/// Aliases never change the declaration's stable identity.
/// </summary>
public sealed record SemanticAlias
{
    /// <summary>The alias name.</summary>
    public string Name { get; }

    /// <summary>
    /// Optional evidence weight in the inclusive range 1-100. When
    /// null, the alias carries no weight and is not considered by weight-based
    /// evidence gating (see <see cref="AliasWeightEvidenceGate"/>).
    /// </summary>
    public int? Weight { get; }

    public SemanticAlias(string name, int? weight = null)
    {
        if (weight is < 1 or > 100)
        {
            throw new ArgumentOutOfRangeException(
                nameof(weight),
                weight,
                "Alias weight must be between 1 and 100 (inclusive) when specified.");
        }

        Name = name;
        Weight = weight;
    }

    public override string ToString() => Name;
}
